File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        target = 0
        for ch in s1:
            target += hash(ch)
        N = len(s1)
        check = 0
        if N > len(s2): return False
        for ch in s2[:N]:
            check += hash(ch)
        if check == target: return True
        l = 0
        for ch in s2[N:]:
            check += hash(ch) - hash(s2[l])
            l += 1
            if check == target:
                return True
        return False

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int count = 0, l = 0, len = s1.length();
        char ch;
        vector <int> chars1(128);
        for(auto ch: s1){
            chars1[ch] += 1;
        }
        vector <int> chars2(128);
        for(int i = 0;i < s2.length(); ++i){
            ch = s2[i];
            if (chars1[ch] != 0){
                if(chars2[ch] < chars1[ch]){
                    ++count;
                    ++chars2[ch];
                    if (count == len) return true;
                }
                else{
                    while(s2[l] != ch){
                        --chars2[s2[l]];
                        --count;
                        ++l;
                    }
                    ++l;
                }
            }
            else{
                chars2 = vector <int>(128);
                count = 0;
                l = i + 1;
            }
        }
        return false;
    }
};